### PR TITLE
run only modified spec

### PR DESCRIPTION
**Pull request by lemonowl**
The PR attempts to address the issue of inefficient test runs.  

In this implementation, the test pool is determined based on the directories where changes were made:  
- If changes are in the framework folders, all tests are executed.  
- If changes are only in the specs or scripts folders, a list of modified specs is compiled and executed.

Full spec runs on release versions have been moved to Jenkins.

Run with changes in spec
Снимок экрана от 2024-12-09 11-03-18

Run without changes
Снимок экрана от 2024-12-09 10-57-55

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -30,3 +30,24 @@ desc 'run project spec'
...
+    # get changes in scripts and find them in spec
+    scripts_diff = `git diff --name-only origin/master -- js python | xargs -I {} grep -Rl {} spec`
+    # get changes in spec
+    spec_diff = `git diff --name-only origin/master -- spec ':!spec/spec_helper.rb' ':!spec/test_data.rb'`
+    files = spec_diff.split | scripts_diff.split
+    if files.all? { |element| element =~ %r{^spec/.*\.rb} }
+      files.empty? ? print('NO TESTS TO RUN.') : sh("bundle exec parallel_rspec #{files.join(' ')}")
+    else
+      print("An incorrect file type for rspec has been detected: #{files}")
+      Rake::Task['default'].invoke
```
Comment (VitaliySerov) [inline Rakefile:48]: Both this line and line 51 calls `Rake::Task['default'].invoke` - probably can be optimized in one single if statement with implications of decision tree

But just a minor complaining, overwise I think it's good, but need to see it in long run if this solution is solve the problem good

File: Rakefile [hunk 2]
```diff
@@ -30,3 +30,24 @@ desc 'run project spec'
...
   system('bundle exec parallel_rspec project_spec')
 end
+
+desc 'run tests in modified specs'
+task :in_modified_specs do
+  # get changes in framework
+  lib_diff = `git diff --name-only origin/master -- lib dockerfiles Dockerfile`
+  if lib_diff.empty?
+    # get changes in scripts and find them in spec
+    scripts_diff = `git diff --name-only origin/master -- js python | xargs -I {} grep -Rl {} spec`
```
Comment (VitaliySerov) [inline Rakefile:40]: Maybe we can somehow do not use python here (maybe some bash magic)
So python will not required for this script

**Review by VitaliySerov — APPROVED** [review]

File: Rakefile (hunk 2 above)
Comment (lemonowl) [reply Rakefile:40]: This is not the use of Python, it's just a folder name where we store scripts for testing python wrapper

**Review by alinasafonova — APPROVED** [review]

File: Rakefile (hunk 2 above)
Comment (VitaliySerov) [reply Rakefile:40]: Ah, got it, my bad :)

**Review by askonev — APPROVED** [review]

**Pull request merged**